src/lib/optimizer/strategy/like_replacement_rule.cpp

```diff
@@ -0,0 +1,87 @@
+#include "like_replacement_rule.hpp"
+
+#include <algorithm>
+#include <iostream>
+
+#include "all_parameter_variant.hpp"
+#include "constant_mappings.hpp"
+#include "expression/expression_functional.hpp"
+#include "logical_query_plan/abstract_lqp_node.hpp"
+#include "logical_query_plan/lqp_utils.hpp"
+#include "logical_query_plan/predicate_node.hpp"
+#include "logical_query_plan/stored_table_node.hpp"
+#include "operators/operator_scan_predicate.hpp"
+#include "statistics/chunk_statistics/chunk_statistics.hpp"
+#include "storage/storage_manager.hpp"
+#include "storage/table.hpp"
+#include "utils/assert.hpp"
+
+using namespace opossum::expression_functional;  // NOLINT
+
+namespace opossum {
+
+std::string LikeReplacementRule::name() const { return "Like Replacement Rule"; }
+
+void LikeReplacementRule::apply_to(const std::shared_ptr<AbstractLQPNode>& node) const {
+  const auto predicate_node = std::dynamic_pointer_cast<PredicateNode>(node);
+  // Continue only on predicate nodes
+  if (predicate_node == nullptr) {
+    _apply_to_inputs(node);
+    return;
+  }
+  const auto expression = std::dynamic_pointer_cast<BinaryPredicateExpression>(predicate_node->predicate());
+  // Continue only if expressions predicate condition is of type PredicateCondition::Like
+  if (expression == nullptr || expression->predicate_condition != PredicateCondition::Like) {
+    _apply_to_inputs(node);
+    return;
+  }
+  const auto column_expression = std::dynamic_pointer_cast<LQPColumnExpression>(expression->left_operand());
+  const auto value_expression = std::dynamic_pointer_cast<ValueExpression>(expression->right_operand());
+  // Continue only if left expressions is column and right expression is value (column LIKE "asdf%")
+  if (column_expression == nullptr || value_expression == nullptr) {
+    _apply_to_inputs(node);
+    return;
+  }
+  const auto value = boost::get<std::string>(value_expression->value);
+  // Filter strings containing "_", as they cannot be reformulated using this approach
+  if (value.find_first_of("_") != std::string::npos) {
+    _apply_to_inputs(node);
+    return;
+  }
+  const auto offset = value.find_first_of("%");
+
+  // Continue only if the string has a % wildcard and ends with the % wildcard ("asdf%")
+  if (offset == std::string::npos || offset <= 0 || value.length() != offset + 1) {
+    _apply_to_inputs(node);
+    return;
+  }
+  // Calculate lower and upper bound of the search pattern
+  const auto lower_bound = value.substr(0, offset);
+  const auto current_character_value = static_cast<int>(lower_bound.at(lower_bound.length() - 1));
+  // Find next value according to ASCII-table
+  if (current_character_value > 0 && current_character_value < 127) {
+    const auto next_character = static_cast<char>(current_character_value + 1);
+    const auto upper_bound = lower_bound.substr(0, offset - 1) + next_character;
+    const auto lower_bound_node = PredicateNode::make(std::make_shared<BinaryPredicateExpression>(
```

Comment:
Please add a comment why you cannot just emit a BETWEEN here. (and a TODO that we soon might after #1481 )

Reply:
We decided not to rewrite like expressions to between expressions. There are two different rules.
Consider this query part: `column LIKE "a%" AND column <= "asdf"`
If we would reformulate the like to a between in this rule, the outcome would be:
`column BETWEEN "a" AND "b" (EX)  AND column <= "asdf"`
If we have two different rules applied in sequence, the outcome is:
`column BETWEEN "a" AND "asdf"`

Reply:
By "comment" I meant "in the code", sorry to be unclear. I am fine with not generating a BETWEEN for now, but in general: Rules should not know about each other. A BETWEEN scan is faster than two separate scans, so not emitting a *faster* predicate because another rule cannot handle it in an obscure case is a weak argument.

Reply:
@mrks @Bensk1 What is your opinion?

Reply:
Moritz is right. Ideally, the BETWEEN rule would also be able to incorporate other betweens when calculating the minimal necessary between range. I think both are fine for now since it should really be an edge case.

Either way, we should put a comment on that and/or open an issue for improvement to either incorporate existing betweens in the other rule or if you decide to keep it as is that instead a between could be emitted and betweens should be incorporated by the between rule.

Reply:
BETWEEN, apart from being faster to execute, can also be used to prune entire Chunks. Two separate scans are less likely to result in a pruned Chunk. I think we want BETWEEN here.

Reply:
Looking at this again, I'd vote for BETWEEN, too. If we want to cover that edge case where we have another limiting predicate on the same column, that should be the job of the between rule, not this one.